monai/apps/reconstruction/ssim_loss.py

```diff
@@ -0,0 +1,73 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+from monai.utils.type_conversion import convert_to_tensor
+
+
+class SSIM(nn.Module):
+    """
+    Build a Pytorch version of the SSIM loss function based on the original formula of SSIM
+
+    Modified and adopted from:
+        https://github.com/facebookresearch/fastMRI/blob/main/banding_removal/fastmri/ssim_loss_mixin.py
+    """
+
+    def __init__(self, win_size: int = 7, k1: float = 0.01, k2: float = 0.03):
+        super().__init__()
+        self.win_size = win_size
+        self.k1, self.k2 = k1, k2
+        self.register_buffer("w", torch.ones(1, 1, win_size, win_size) / win_size**2)
+        np = win_size**2
+        self.cov_norm = np / (np - 1)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor, data_range: torch.Tensor) -> torch.Tensor:
+        """
+        y and x are of shape (B,C,H,W) where B is batch_size, C is number of channels, and (H,W) are height and width.
+
+        Args:
+            x: first sample (e.g., the reference image).
+            y: second sample (e.g., the reconstructed image)
+            data_range: dynamic range of the data
+
+        Returns:
+            1-ssim_score (recall this is meant to be a loss function)
+
+        Example:
+            .. code-block:: python
+
+                import torch
+                x = torch.ones([1,1,10,10])/2
+                y = torch.ones([1,1,10,10])/2
+                data_range = x.max().unsqueeze(0)
+                # the following line should print 1.0 (or 0.9999)
+                print(1-SSIM()(x,y,data_range))
+        """
+        data_range = data_range[:, None, None, None]
+        c1 = (self.k1 * data_range) ** 2
```

Comment:
Could you change variable names to make them more readable? Thank you!

Reply:
Please avoid using a1, b1, c1 as variable names. Thanks!

Reply:
ah, c1 and c2 are standard ssim coefs (see [here](https://medium.com/srm-mic/all-about-structural-similarity-index-ssim-theory-code-in-pytorch-6551b455541e) for instance)... But a's and b's are not so I'll change them.